### PR TITLE
Authentication audit logging with authentication mechanism

### LINKED ISSUE #82295: Authentication audit logging - denote the authentication mechanism used.
**What would you like to be added**:

Logging for the authentication mechanism used by a user for requests to the API server.

**Why is this needed**:

At the moment Kubernetes does not put the mechanism used to authenticate a user into it's audit logs.  As Kubernetes supports multiple authentication mechanisms, this could lead to a circumstance where an identical username is defined under different authentication schemes and it would be impossible to identify which had been used for a given request.

This is particularly serious in the case of client certificate authentication. As all that is required for the creation of client certificate credentials is access to the `ca.key` file for the cluster and credentials can be created using `openssl` commands, there may be no audit trail of users created with this mechanism.  

An attacker who gained read-only access to that file would be able to create new credentials with the same usernames as users authenticated via other mechanisms, removing the ability of cluster operators to accurately audit user actions.

### PR DESCRIPTION
**Pull request by tedyu**
**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:
As #82295 requested, this PR adds audit logging with authentication mechanism.

**Which issue(s) this PR fixes**:
Fixes #82295

```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

File: staging/src/k8s.io/apiserver/pkg/endpoints/filters/authentication.go [hunk 1]
```diff
@@ -42,9 +44,10 @@ import (
  * the metric stability policy.
  */
 const (
-	successLabel = "success"
-	failureLabel = "failure"
-	errorLabel   = "error"
+	successLabel  = "success"
+	failureLabel  = "failure"
+	errorLabel    = "error"
+	annotationKey = "authentication.k8s.io/auth-mechanism"
```
Comment (tallclair) [inline staging/src/k8s.io/apiserver/pkg/endpoints/filters/authentication.go]: Maybe `authentication.k8s.io/authenticator-name`?

File: staging/src/k8s.io/apiserver/pkg/endpoints/filters/authentication.go [hunk 2]
```diff
@@ -98,6 +101,9 @@ func WithAuthentication(handler http.Handler, auth authenticator.Request, failed
 		if len(apiAuds) > 0 {
 			req = req.WithContext(authenticator.WithAudiences(req.Context(), apiAuds))
 		}
+		ctx := req.Context()
+		ae := request.AuditEventFrom(ctx)
```
Comment (tallclair) [inline staging/src/k8s.io/apiserver/pkg/endpoints/filters/authentication.go]: Nit: prefer inlining
```suggestion
		ae := request.AuditEventFrom(req.Context())
```

File: staging/src/k8s.io/apiserver/pkg/endpoints/filters/authentication.go (hunk 1 above)
Comment (tallclair) [inline staging/src/k8s.io/apiserver/pkg/endpoints/filters/authentication.go]:
```suggestion
	auditAnnotationKey = "authentication.k8s.io/auth-mechanism"
```

File: staging/src/k8s.io/apiserver/pkg/endpoints/filters/authentication.go [hunk 3]
```diff
@@ -98,6 +101,9 @@ func WithAuthentication(handler http.Handler, auth authenticator.Request, failed
...
+		audit.LogAnnotation(ae, annotationKey, getAuthMethods(req))
```
Comment (tallclair) [inline staging/src/k8s.io/apiserver/pkg/endpoints/filters/authentication.go]: This method is just for providing a failure message, it doesn't actually give the method used.

**Review by tallclair — COMMENTED** [review]
Needs a unit test, which would highlight the incorrect use of `getAuthMethods`

File: staging/src/k8s.io/apiserver/pkg/endpoints/filters/authentication.go (hunk 3 above)
Comment (tedyu) [reply staging/src/k8s.io/apiserver/pkg/endpoints/filters/authentication.go]: I am looking at 
```
func (w *WebhookTokenAuthenticator) AuthenticateToken(ctx context.Context, token string) (*authenticator.Response, bool, error) {
```
to see if useful information for audit can be obtained.

**Pull request closed without merging**